### ISSUE TITLE
False positive for never_loop with complex nested control flow

### ISSUE THREAD

**Issue by Nemo157** [rust-lang/rust-clippy#4058]
```rust
fn main() {
    let mut n = 1;

    let n = loop {
        break match n {
            5 => n,
            _ => {
                n += 1;
                continue;
            }
        }
    };

    println!("{}", n);
}
```

([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8b23dd141a92648c3d76647ede87e391)) gives

```
error: this loop never actually loops
  --> src/main.rs:4:13
   |
4  |       let n = loop {
   |  _____________^
5  | |         break match n {
6  | |             5 => n,
7  | |             _ => {
...  |
11 | |         }
12 | |     };
   | |_____^
   |
   = note: #[deny(clippy::never_loop)] on by default
```

but running it will count `n` up to 5 before breaking and returning the final value.

**Comment by flip1995** [comment]
This is definitely a bug. But in addition to fixing this, I would write a style lint, which prevents `continue`ing out of a `break` expression. Couldn't you write this as 

```rust
match n {
    5 => break n,
    _ => {
        n += 1;
        continue;
    }
}
```

I'm sure the code where you encountered this FP is more complex, but IMO it is not really obvious, what this code does. If it's more complex code, this is probably even more confusing. What do you think?

**Comment by Nemo157** [comment]
In the [real code](https://github.com/rustasync/async-compression-rs/pull/11/files#diff-2c8464bd9cd2433d01773772ee45acb2R59) there’s six branches that return a value to break with and two branch that update some state and continue the loop. I find it cleaner to have it such that just finishing the match with a value returns it and the few reruns are `continue` instead of having 6 early returns.

**Comment by flip1995** [comment]
Yeah I can see why some may prefer that instead of spreading breaks.

This should be fixable by walking the break expression, searching for a `continue` expression (and don't emit the Lint in that case).

**Comment by udoprog** [comment]
So this might possibly be related.

I have a number of `loop` which gets (correctly) flagged as `clippy::never_loop`, but they are strictly speaking not useless, but instead uses break-with-value patterns, like this (parses '0x', '0b', '0o' prefixes):

```rust
let number = {
    let mut sub = it.clone();

    #[allow(clippy::never_loop)]
    loop {
        let m = match (sub.next(), sub.next()) {
            (Some((_, '0')), Some((_, m))) => m,
            _ => break LitNumber::Decimal,
        };

        let number = match m {
            'x' => LitNumber::Hex,
            'b' => LitNumber::Binary,
            'o' => LitNumber::Octal,
            _ => break LitNumber::Decimal,
        };

        // consume two character.
        it.next();
        it.next();
        break number;
    }
};
```

These do not purely "obfuscate" as proposed by the lint, but actually serves a control flow purpose. So it might be worth considering if the lint really should catch loops which have multiple branches of `break`'s.

I have six more examples like these in the project, and some of them will indeed be rewritten, others are harder to rewrite cleanly. Usually it involves [complex chains of matching with a fallback that errors](https://github.com/udoprog/runestick/blob/801c3644e04fcd42557b25d75613a7f517e6d5a7/crates/rune/src/compiler/mod.rs#L932).
